// @react-leaflet/core/cjs/layer.js
"use strict";

exports.__esModule = true;
exports.useAttribution = useAttribution;
exports.useLayerLifecycle = useLayerLifecycle;
exports.createLayerHook = createLayerHook;

var _react = require("react");

var _context = require("./context");

var _events = require("./events");

function useAttribution(map, attribution) {
  const attributionRef = (0, _react.useRef)(attribution);
  (0, _react.useEffect)(function updateAttribution() {
    if (attribution !== attributionRef.current && map.attributionControl != null) {
      if (attributionRef.current != null) {
        map.attributionControl.removeAttribution(attributionRef.current);
      }

      if (attribution != null) {
        map.attributionControl.addAttribution(attribution);
      }
    }

    attributionRef.current = attribution;
  }, [map, attribution]);
}

function useLayerLifecycle(element, context) {
  (0, _react.useEffect)(function addLayer() {
    var _context$layerContain;

    const container = (_context$layerContain = context.layerContainer) != null ? _context$layerContain : context.map;
    container.addLayer(element.instance);
    return function removeLayer() {
      container.removeLayer(element.instance);
    };
  }, [context, element]);
}

function createLayerHook(useElement) {
  return function useLayer(props) {
    const context = (0, _context.useLeafletContext)();
    const elementRef = useElement(props, context);
    useAttribution(context.map, props.attribution);
    (0, _events.useEventHandlers)(elementRef.current, props.eventHandlers);
    useLayerLifecycle(elementRef.current, context);
    return elementRef;
  };
}
